reference/src/functions/macros.md

```diff
@@ -0,0 +1,543 @@
+substituted by expression. In addition, the caller can supply code to the macro via
+[lambdas](#lambdas).
+
+These expression substitution mechanics make `macro` functions similar
+[to macros found in other programming languages](<https://en.wikipedia.org/wiki/Macro_(computer_science)>);
+however, they are more constrained in Move than you might expect from other languages. The parametrs
+and return values of `macro` functions are still typed--though this can be partially relaxed with
+the [`_` type](TODO). The upside of this restriction however, is that `macro` functions can be used
+anywhere a normal function can be used, which is notably helpful with
+[method syntax](../method-syntax.md).
+
+A more extensive
+[syntactic macro](<https://en.wikipedia.org/wiki/Macro_(computer_science)#Syntactic_macros>) system
+may come in the future.
+
+## Syntax
+
+`macro` functions have a similar syntax to normal functions. However, all type parameter names and
+all parameter names must start with a `$`.
+
+```text
+<visibility>? macro fun <identifier><[$type_parameters: constraint],*>([$identifier: type],*): <return_type> <function_body>
+```
+
+For example, the following `macro` function takes a vector and a lambda, and applies the lambda to
+each element of the vector to construct a new vector.
+
+```move
+macro fun map<$T, $U>($v: vector<$T>, $f: |$T| -> $U): vector<$U> {
+    let mut v = $v;
+    v.reverse();
+    let mut i = 0;
+    let mut result = vector[];
+    while (!v.is_empty()) {
+        result.push_back($f(v.pop_back()));
+        i = i + 1;
+    };
+    result
+}
+```
+
+The `$` is there to indicate that these do not behave like their normal, non-macro counterparts. For
+type parameters, they can be instantiated with any type (even a reference type `&` or `&mut`), and
+they will satisfy any constraint. Similarly for parameters, they will not be evaluated eagerly, and
+instead the argument expression will be substituted at each usage.
+
+## Lambdas
+
+Lambdas a new type of expression that can be used only with `macro`s. These are used to pass code
+from the caller into the body of the `macro`. While the substition is done at compile time, they are
+used similarly to [anonymous functions](https://en.wikipedia.org/wiki/Anonymous_function),
+[lambdas](https://en.wikipedia.org/wiki/Lambda_calculus), or
+[closures](<https://en.wikipedia.org/wiki/Closure_(computer_programming)>) in other languages.
+
+As seen in the example above (`$f: |$T| -> $U`), lambdas types defined with the syntax
+
+```text
+|<type>,*| (-> <type>)?
+```
+
+Lambda expressions then are defined at the call site of the `macro` the syntax
+
+```text
+|(<identifier> (: <type>)?),*| <expression>
+|(<identifier> (: <type>)?),*| -> <type> { <expression> }
+```
+
+Note that if the return type is annotated, the body of the lambda must be enclosed in `{}`.
+
+Using the `map` macro defined above
+
+```move
```

Comment:
Will macros permit receiver syntax (now or ever)? If so, should we add an example of that here?

Reply:
I need to add a section yeah, especially since it changes things a bit for evaluation order